std/range/package.d

```diff
@@ -9213,3 +9221,293 @@ if (is(typeof(fun) == void) || isSomeFunction!fun)
+
+    static immutable r1 = [1, 2, 3, 4];
+    static immutable r2 = [0, 0, 1, 2, 3, 4];
+    assert(r1.padLeft(0, 6).equal(r2));
+}
+
+/**
+Extend the length of the input range $(D r) by padding out the end of the range
+with the element $(D e). The element $(D e) must be of a common type with the
+element type of the range $(D r) as defined by $(REF CommonType, std, traits).
+If $(D n) is less than the length of of $(D r), then the contents of $(D r) are
+returned.
+
+The range primitives that the resulting range provides depends whether or not $(D r)
+provides it. Except the functions $(D back) and $(D popBack), which also require
+the range to have a length as well as $(D back) and $(D popBack).
+
+Params:
+    r = an input range with a length
+    e = element to pad the range with
+    n = the length to pad to
+
+Return:
+    a range containing the elements of the original range with the extra padding
+*/
+auto padRight(R, E)(R r, E e, size_t n) if (
+    isInputRange!R &&
+    !isInfinite!R &&
+    !is(CommonType!(ElementType!R, E) == void))
+{
+    static struct Result
+    {
+        private:
+        R data;
+        E element;
+        size_t counter;
+        static if (isBidirectionalRange!R && hasLength!R) size_t backPosition;
+        size_t maxSize;
+
+        public:
+        bool empty() @property
+        {
+            return data.empty && counter >= maxSize;
+        }
+
+        auto front() @property
+        {
+            return (data.empty) ? element : data.front;
+        }
+
+        void popFront()
+        {
+            ++counter;
+
+            if (!data.empty)
+            {
+                data.popFront;
+            }
+        }
+
+        static if (hasLength!R)
+        {
+            size_t length() const @property
+            {
+                return data.length > maxSize ? data.length : maxSize;
+            }
+        }
+
+        static if (isForwardRange!R)
+        {
+            auto save() @property
+            {
+                typeof(this) result = this;
+                data = data.save;
+                return result;
+            }
+        }
+
+        static if (isBidirectionalRange!R && hasLength!R)
+        {
+            auto back() @property
+            {
+                return (backPosition > data.length) ? element : data.back;
+            }
+
+            void popBack()
+            {
+                if (backPosition > data.length)
+                {
+                    --backPosition;
+                    --maxSize;
+                }
+                else
+                {
+                    data.popBack;
+                }
+            }
+        }
+
+        static if (isRandomAccessRange!R)
+        {
+            inout(E) opIndex(size_t index) inout
+            {
+                if (index > data.length && index <= maxSize)
+                {
+                    return element;
+                }
+
+                return data[index];
```

Comment:
Only for arrays. For general ranges, it's probably a bad idea to assume that it's OK to pass in an invalid index. At the very least we should have an `assert` that the index is within bounds, so that it behaves like an array bounds check.